core/src/test/java/io/grpc/AbstractForwardingTest.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import static junit.framework.TestCase.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mockingDetails;
+import static org.mockito.Mockito.verify;
+
+import com.google.common.base.Defaults;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * An abstract base class for testing forwarding classes. This automatically checks that
+ * all public methods of the {@link #delegateClass()} are forwarded by the forwarder to the
+ * delegate. This does NOT verify that arguments are forwarded properly. It only alerts
+ * the developer if a forward method is missing.
+ */
+@RunWith(JUnit4.class)
+public abstract class AbstractForwardingTest<T> {
+  /**
+   * Returns a mock object that can be used with {@code Mockito.verify()}.
+   */
+  public abstract T mockDelegate();
+
+  /**
+   * Returns a forwarder object that wraps around the object returned by {@link #mockDelegate()}.
+   */
+  public abstract T forwarder();
+
+  /**
+   * Returns the class of the object being forwarded. There is no easy way to find this
+   * reflectively from the value of {@link #mockDelegate()}.
+   * {@code mockDelegate().getClass().getSuperclass()} works at the moment but this would rely on
+   * mockito internals. It's easier to explicitly specify the class.
+   */
+  public abstract Class<T> delegateClass();
+
+  @Test
+  public void methodsForwarded() throws Exception {
```

Comment:
nit: I think this should be a protected final method, and accept a set of methods to ignore.  The base class can call this test.  Sometimes a final method or a constructor (like Builder.build()) messes up this reflection test.

Reply:
I think `@Test` can only be `public`.

Reply:
How about an `@Override`able method that returns a list of ignored methods? This way we minimize the risk that a developer subclass this test and fail to call the test method, but see that all tests are green.

Alternatively, make all these things be constructor args to the class. It may make initializing slightly annoying because it forces everything to be created by the time we call `super()`.

Reply:
I'm slightly against putting tests in the parent class, mainly because it is annoying to run them manually.  You can't click on the super class method to run from the child class.  

As for people forgetting: if the super class became a utility class there would be no risk of forgetting.

Reply:
At least in intellij, right clicking the superclass method lets you choose running a child class or running all child classes.